src/main/java/bsh/BshMethod.java

```diff
@@ -345,6 +345,11 @@ private Object invokeImpl(
         else
             localNameSpace = new NameSpace( declaringNameSpace, name );
 
+        // This look dubious because in the case of overrideNameSpace
+        // because will clobber the previous value.  OverrideNameSpace==true
+        // is only used in this codebase by PreparsedScripts and there it is
+        // safe but any other usage might cause problems.  This should be
+        // cleaned up.It was changed as part of #508 but that fix was wrong.
         localNameSpace.isMethod = true;
```

Comment:
You did not explain why this fix is wrong, was asked to elaborate please.

We are in BshMethod.invokeImpl so these are all methods. Cannot be anything else and never will be. So they should all be marked as isMethod = true. 

How do you fix 508 then?

Reply:
I really should have made this remark as a code review comments and not actually put it in the code.  My bad.

You are correct that always these should be marked as isMethod = true in this place.

My concern was that when you are using the cache that you would get a namespace object and mark it here as a method.  Later on when the object gets recycled it may be used for a non-method block, but the method flag has been set here.  There is no place where the flag gets unset even though the namespace is being recycled.

Reply:
You were right, it was making a real mess there, not even sure why it resolved the issue either. Even the global namespace was marked as a method. Well spotted on the cached namespaces, it is not enough for unique name spaces to only use the parent namespace. Subsequent blocks are bungled up and yes inherit the method flag too. This certainly was not the intention...

I have tucked this flag set back in under the else where it was and added a unique block id to the block name space cache, per block instance which now appears to behave like it should. Creates a new namespace per unique block and only reuses it for that specific block, if needed for recursion, loops, etc. The performance still seems on par with the original implementation and the namespaces are not being polluted anymore.

I tried to merged all your other fixes in from this branch, will have to see if I missed anything.